### PR TITLE
Enhance codebase with PEP 484 compliance and performance optimizations

### PR DESCRIPTION
**Pull request by flferretti**
This pull request updates the codebase to comply with PEP 484 for type hints. Key improvements include:

- Enhanced exception handling.
- Removed unused tree transforms and imports.
- Eliminated assignments to methods.
- Replaced dict and list literals with safer alternatives.
- Reduced memory footprint during robot dict creation.
- Adopted f-strings where applicable.

These changes improve code readability, maintainability, and overall adherence to Python best practices.

### CODE REVIEW

File: src/rod/sdf/common.py [hunk 1]
```diff
@@ -89,7 +89,9 @@ def transform(self) -> npt.NDArray:
         )
 
     @staticmethod
-    def from_transform(transform: npt.NDArray, relative_to: str = None) -> "Pose":
+    def from_transform(
+        transform: npt.NDArray, relative_to: str | None = None
+    ) -> "Pose":
```
Comment (diegoferigo) [inline src/rod/sdf/common.py:94]: Here you can also import annotations and use update the returned type

**Review by diegoferigo — CHANGES_REQUESTED** [review]
Thanks for this PR @flferretti, it includes a polishing that was long awaited. Some comment:

- All the typing and caching modifications are more than welcome.
- I'm concerned about the change of methods like `links()|joints()|etc` to attributes. It's an API change that would break all the downstream code and I'm strongly against it. The idea of all the methods of these classes is that they provide some kind of post-processing of the parsed SDF. In this specific case, it allows downstream code to operate always on lists. Keep in mind that this need is due to the fact that the XML is parsed from an external library, and it populates the attributes as object of list of objects depending on the XML itself.
- Please double check all the cached properties. For example, the proposed `Model.links` cached property (change that I do not support as explained) might create problems because the `Model` is mutable, and users can add additional links after creation. The cache would return the wrong list of links in this case.

File: src/rod/sdf/common.py (hunk 1 above)
Comment (flferretti) [reply src/rod/sdf/common.py]: Done in https://github.com/ami-iit/rod/pull/39/commits/137415fdccb4680ee4cbe0bf971e97338ae87fa1

**Review by diegoferigo — APPROVED** [review]
Perfect, it looks good now, thank you @flferretti.

**Pull request merged**